Tiered Storage manual [HZ-4383] (#1047)
hazelcast/hz-docs: docs/modules/tiered-storage/pages/overview.adoc

This data structure spans the memory and the configured disk to provide the aggregated capacity of the resources available for storing IMap's entries.
There is a HybridLog data structure for every partition of every IMap.

The IMap entries stored in these HybridLog instances are internally referenced by their logical addresses.
The key to logical address mapping uses a hash index component.
These hash indices are hashtables using separate chaining for conflict resolution.
The table of these hash indices is stored in native memory; the chaining information for the buckets is stored on the disk.

Tiered Storage ensures that every accessed IMap entry is available in the memory to build a hot cache of entries.
This means that, if an entry is on the disk, it is relocated to the memory inside the HybridLog hosting the entry, and its previous location on the disk becomes unreachable.
As more and more entries are relocated to the in-memory region of the HybridLog, the memory space occupied by the unreachable entries on disk is reclaimed.
This reclamation is done by a compactor running synchronously with IMap operations on the same partition.

== Example Use Case

> Efficiently managing transactional data is crucial in banking applications. Traditional setups like involving memory and a database can be complex and costly to maintain.

Author: I'm afraid this statement may put a target on us. It might give the reader the impression that tiered storage deals with data with strong consistency guarantees, while it's not the case. What's [written](https://docs.hazelcast.com/hazelcast/5.3/data-structures/map#availability-vs-consistency) for distributed maps still holds. I'd either reference that page, or make this remark in any way, or remove this example.

{description}

NOTE: If your Enterprise license was generated before Hazelcast Platform version 5.2, you'll need a new Enterprise license that enables Tiered Storage.
See xref:deploy:enterprise-licenses.adoc[].

Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API.
It allows you to optimize infrastructure costs and performance by combining a fast, high performance in-memory layer and a lower-cost, high-capacity disk-based storage layer.

> With a larger data capacity per member, Tiered Storage can also reduce the number of deployed Hazelcast cluster members.

Reviewer: ```suggestion
With a larger data capacity per member, Tiered Storage can also reduce the number of deployed Hazelcast cluster members.
```
Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API. It allows you to optimize infrastructure costs and performance by combining a fast in-memory layer with a cost-effective, high-capacity disk-based storage layer. 

Furthermore, with a extended data capacity per member, Tiered Storage can also reduce the number of deployed Hazelcast cluster members.

Author: Addressed in https://github.com/hazelcast/hz-docs/pull/1047/commits/85eb21450d0833a0e33da49a49d306678275a0ac.

> Accessing data in the memory offers superior access speed, but memory is much more expensive than disks, while the modern NVMe SSDs offer performance that makes Tiered Storage a viable alternative to memory-only storage at a much lower total cost of ownership.

Reviewer: Im not convinced this adds additional value to what has already been discussed in the previous paragraph.

Author: Removed in https://github.com/hazelcast/hz-docs/pull/1047/commits/85eb21450d0833a0e33da49a49d306678275a0ac.

{description}

NOTE: If your Enterprise license was generated before Hazelcast Platform version 5.2, you'll need a new Enterprise license that enables Tiered Storage.
See xref:deploy:enterprise-licenses.adoc[].

Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API.
It allows you to optimize infrastructure costs and performance by combining a fast, high performance in-memory layer and a lower-cost, high-capacity disk-based storage layer.

> Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API. It allows you to optimize infrastructure costs and performance by combining a fast in-memory layer with a cost-effective, high-capacity disk-based storage layer.

Reviewer 2: >Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API. It allows you to optimize infrastructure costs and performance by combining a fast in-memory layer with a cost-effective, high-capacity disk-based storage layer.

This part seems duplicate information. Previous paragraph also has same info.

Author: Removed the duplicate.

> 

Author: Yes, we release it as GA.

{description}

NOTE: If your Enterprise license was generated before Hazelcast Platform version 5.2, you'll need a new Enterprise license that enables Tiered Storage.
See xref:deploy:enterprise-licenses.adoc[].

== What Can You Do with Tiered Storage?

> Have fewer Hazelcast members for large maps. For example, for a map with 1 TB of data, you'll need

Reviewer: Suggestion: Remove 'what can you do with TS' and amalgamate these two sections.

Tiered Storage allows you to work with datasets that are larger than the available memory using the same Hazelcast map API. It allows you to reduce infrastructure costs by utilizing cheaper disk storage and reducing the number of Hazelcast cluster members deployed.

Author: Made an attempt in 06a2a152059bfb69e1201dfdbf7842c9b65f589f, please take a look if that's what you meant.

ten Hazelcast members if all data is kept in memory and each member has 100 GB of data. Tiered Storage allows you to split the data between memory and disks, which requires fewer cluster members depending on how the data is distributed.

== When to Use Tiered Storage

Accessing data in the memory offers superior access speed, but memory is much more expensive than disks, while the modern NVMe SSDs offer performance that makes Tiered Storage a viable alternative to memory-only storage at a much lower total cost of ownership.

> Consider using Tiered Storage if one of the following is true for your use-case.

Reviewer: An additional use case to consider:
You have significant temporal access patterns inside a larger dataset

Author: Added in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

Reviewer: Minor update: 
You have significant temporal access patterns inside a very large dataset.

I would also move this up to be positioned below the bullet 'the hot set of entries...'

- The data size Hazelcast manages exceeds the capacity of the memory.

> - You will to trade off performance for lower ownership costs.

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

{description}

NOTE: If your Enterprise license was generated before Hazelcast Platform version 5.2, you'll need a new Enterprise license that enables Tiered Storage.
See xref:deploy:enterprise-licenses.adoc[].

Tiered Storage allows you to work with datasets that are larger than memory using the same Hazelcast map API. It allows you to optimize infrastructure costs and performance by combining a fast in-memory layer with a cost-effective, high-capacity disk-based storage layer.

Furthermore, with a extended data capacity per member, Tiered Storage can also reduce the number of deployed Hazelcast cluster members.

Consider using Tiered Storage if one of the following is true for your use-case.

- The data size Hazelcast manages exceeds the capacity of the memory.

> - You are willing to trade off performance for lower ownership costs.

Author: It wants to mean you pay less, but the performance will degrade.

> - You are willing to trade off performance for lower ownership costs.

Reviewer 3: ```suggestion
- You want lower ownership costs, despite that having an impact on performance.
```

> - You are willing to trade off performance for lower ownership costs.

Reviewer 3: OK, then my first suggestion ought to work, I misread when I said higher costs, but perhaps the second suggestion is clearer. 

Can we quantify how much degradation is likely to occur? Even as slight, or small (I hope). Or maybe an example of when this is a fair trade?

- Data Persistence and Tiered Storage are mutually exclusive features. If both are enabled simultaneously, `InvalidConfigurationException` is thrown during the member startup.

- SQL is not supported for the Tiered-Storage-backed maps. `UnsupportedOperationException` is thrown when SQL query is executed. To avoid this, use xref:query:predicate-overview.adoc[Predicate API] instead.

== Supported Devices

Hazelcast supports only NVMe (local) SSD devices for Tiered Storage.
While HDDs and some network storages work with the same file abstraction that local SSDs do, and technically speaking such setups work, the access latency of these devices compared to that of the modern SSDs are so high that using Tiered Storage with such devices are not supported.

== High-Level Design Overview

> The Tiered Storage feature for IMaps is backed by a special-purpose, addressable data structure called HybridLog.

Reviewer 2: ```suggestion
The Tiered Storage feature for IMaps is backed by a special-purpose, addressable data structure called HybridLog.
It is called HybridLog because it seamlessly combines in-place updates with a traditional append-only log.
```

Author: Added in https://github.com/hazelcast/hz-docs/pull/1047/commits/0253a16cd6c5ede529c5f0f74b0e730fe62b5d95

- Data Persistence and Tiered Storage are mutually exclusive features. If both are enabled simultaneously, `InvalidConfigurationException` is thrown during the member startup.

- SQL is not supported for the Tiered-Storage-backed maps. `UnsupportedOperationException` is thrown when SQL query is executed. To avoid this, use xref:query:predicate-overview.adoc[Predicate API] instead.

== Supported Devices

Hazelcast supports only NVMe (local) SSD devices for Tiered Storage.

> While HDDs and some network storages work with the same file abstraction that local SSDs do, and technically speaking such setups work, the access latency of these devices compared to that of the modern SSDs are so high that using Tiered Storage with such devices are not supported.

Reviewer: Do we officially want to say we don't support non-SSD devices? Instead, should we recommend the use of SSDs and detail I/O access time considerations/impact when using network devices?

Author: I'd keep support, because we targeting SSDs only, and our tests cover SSDs only. We just can't limit the supported device due to the lack of the information about the device type.

Reviewer: Does Danny test using ESBs?

== High-Level Design Overview

The Tiered Storage feature for IMaps is backed by a special-purpose, addressable data structure called HybridLog.
This data structure spans over the memory and the configured disk making the aggregated capacity of the resources available for storing IMap's entries.
There is a HybridLog data structure for every partition of every IMap.

The IMap entries stored in these HybridLog instances are internally referenced by their logical addresses.
The key to logical address mapping is done by a hash index component.
These hash indices are hashtables using separate chaining for conflict resolution.
The table of these hash indices are stored in the native memory, whereas the chaining information of the buckets are stored on the disk.

Tiered Storage makes sure every accessed IMap entry is available in the memory to build a hot cache of entries.
This means, if an entry is on the disk, first it gets relocated to the memory inside the HybridLog hosting the entry, and its previous location on the disk becomes unreachable.

> As more and more entries get unreachable, to prevent running out of disk space the disk space occupied by these unreachable entries has to be reclaimed.

Reviewer: Suggestion: 
-As more and more entries get unreachable, to prevent running out of disk space the disk space occupied by these unreachable entries has to be reclaimed.- 
As more and more entries are relocated to the in-memory region of the HybridLog, the memory space occupied by the unreachable entries on disk are reclaimed.

Author: Replaced in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

This reclamation is done by a compactor running synchronously with regards to the IMap operations of the same partition.

> The design of Tiered Storage is based on the concepts and data structures introduced by the https://www.microsoft.com/en-us/research/uploads/prod/2018/03/faster-sigmod18.pdf[FASTER: A Concurrent Key-Value Store with In-Place Updates] Microsoft research paper.

Reviewer: How open do we want to be about the internal implementation of TS in this doc?

Author: Good question. We can avoid naming the structure and removing the reference to the paper. But if we ever want to write a blog post about the internals, it's inevitable to mention FASTER (or we will be criticized for not giving credit to its authors).

It's a valid question though if we want our manual referring to external links.

Reviewer 4: We try not to reference papers directly in the product documentation, instead we are referencing to our blogs in case the reader wants to know more in-depth information.

Author: Removed the reference to FASTER in 06a2a152059bfb69e1201dfdbf7842c9b65f589f
